File: 1456. 定长子串中元音的最大数目/main.js
var maxVowels = function (s, k) {
  let ans = 0,
    vowel = 0;
  for (let i = 0; i < s.length; i++) {
    // 枚举窗口右端点 i
    // 1. 右端点进入窗口
    if (
      s[i] === "a" ||
      s[i] === "e" ||
      s[i] === "i" ||
      s[i] === "o" ||
      s[i] === "u"
    ) {
      vowel++;
    }

    const left = i - k + 1; // 窗口左端点
    if (left < 0) {
      // 窗口大小不足 k，尚未形成第一个窗口
      continue;
    }

    // 2. 更新答案
    ans = Math.max(ans, vowel);

    // 3. 左端点离开窗口，为下一个循环做准备
    let out = s[left];
    if (
      out === "a" ||
      out === "e" ||
      out === "i" ||
      out === "o" ||
      out === "u"
    ) {
      vowel--;
    }
  }
  return ans;
};

// 作者：灵茶山艾府
// 链接：https://leetcode.cn/problems/maximum-number-of-vowels-in-a-substring-of-given-length/solutions/2809359/tao-lu-jiao-ni-jie-jue-ding-chang-hua-ch-fzfo/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: 1456. 定长子串中元音的最大数目/main.rs
impl Solution {
    pub fn max_vowels(s: String, k: i32) -> i32 {
        let bytes = s.as_bytes();
        let k = k as usize;
        let mut count = 0;
        let mut ans = 0;

        for i in 0..bytes.len() {
            if Self::is_vowel(bytes[i]) {
                count += 1;
            }
            // 原来的写法里 let left = i - k + 1 使用的是 usize，当 i < k - 1 时会产生下溢（无符号整数不能表示负数），结果变成一个非常大的数，比如错误里看到的 18446744073709551614。接下来访问 bytes[left] 时就越界了
            // let left = i - k + 1;
            if i + 1 < k {
                // 注意这里的判断不能出现负数 usize类型不能出现负数 不然就会越界
                continue;
            }
            let left = i - k + 1;
            ans = ans.max(count);

            if Self::is_vowel(bytes[left]) {
                count -= 1;
            }
        }

        ans
    }

    #[inline]
    fn is_vowel(b: u8) -> bool {
        matches!(b, b'a' | b'e' | b'i' | b'o' | b'u')
    }
}
// https://leetcode.cn/problems/maximum-number-of-vowels-in-a-substring-of-given-length/solutions/2809359/tao-lu-jiao-ni-jie-jue-ding-chang-hua-ch-fzfo/
